src/Components/Web.JS/src/Services/NavigationEnhancement.ts

```diff
@@ -142,6 +156,7 @@ export async function performEnhancedPageLoad(internalDestinationHref: string, f
         } else {
           // For non-get requests, we can't safely re-request, so just treat it as an error
           replaceDocumentWithPlainText(`Error: ${fetchOptions.method} request to ${internalDestinationHref} returned non-HTML content of type ${responseContentType || 'unspecified'}.`);
+          navigationEnhancementCallbacks.documentUpdated();
```

Comment:
I have to admit I'm not really a fan of having *many* different calls to `documentUpdated` from lots of different points in the flow. It looks pretty fragile, like we might either add it somewhere it shouldn't go, or forget to add it in some new place, and then the behavior will be unpredictable.

Ideally we would only need to activate interactive components in a fairly constrained set of places:

 * When the initial document load or a subsequent enhanced navigation is entirely completed, but *only* if there's no ongoing streaming rendering and no other enhanced navigation has started
 * When some streaming rendering completes (as in, entirely completes, not just a single chunk of it)
 * (Potentially) when the WebAssembly runtime completes its async loading process, if that was triggered by having seen a webassembly component marker in one of the two cases above

I don't think we have to deal with cases that are in the middle of a process, since we don't want to activate interactive components until streaming SSR has entirely stopped (because if we did, we wouldn't be able to wait for ComponentState).

I don't think we have to deal with these completely fatal error cases (`replaceDocumentWithPlainText`) either, since there's no going back from these - the page is permanently dead and the user will just have to leave it, which will dispose everything.

What do you think? I appreciate I might be missing some subtleties about why it's been done this way.

Reply:
If the scenarios listed above are accurate, then instead of passing around a collection of callbacks, we could define some concept like a "SSRLoadingState" whereby:

 * Code can call `SSRLoadingState.addWorkItem(somePromise)` to represent a thing that's happening, ideally at the top level of an async process like "perform enhanced nav including streaming SSR"
 * `Boot.Web.ts` or whenever can register a callback like `SSRLoadingState.onLoadingStateChanged(callback)` that fires when we move from 0 pending work items to >= 1, or when we move from >= 1 to 0.

Feel free to ignore this if it's irrelevant. I was just trying to make a constructive suggestion but maybe I'm missing the reasons why it needs notifications at each step in an ongoing process.

Reply:
Update: @MackinnonBuck explained I was missing what's really going on here. `documentUpdated` does *not* scan the document nor activate any new interactive components (that only happens during DOM sync). Instead, `documentUpdated` notifies the root component manager that it may need to update the parameters on existing root components, since we do want those to be updatable at arbitrary points during streaming SSR. However it might be cleaner still to centralize this so that the calls to RootComponentManager happen during DOM sync (since at that point we already know precisely which marker is being updated), and then RootComponentManager can debounce these calls so it only results in a single message to the server. Then we don't need `documentUpdated` at all (or at least, not in these cases).